cfme/generic_objects/definition/ui.py

```diff
@@ -0,0 +1,477 @@
+        description: generic object definition description
+        attributes:  generic object definition attributes ex. {'address': 'string'}
+        associations:  generic object definition associations ex. {'services': 'Service'}
+        methods: generic object definition methods ex. ['method1', 'method2']
+        custom_image_file_path: generic object definition custom image file path
+        cancel: by default will not be canceled, pass True to make successful cancel
+
+    Returns:
+        GenericObjectDefinition entity
+
+    """
+    view = navigate_to(self, 'Add')
+    view.fill({
+        'name': name,
+        'description': description,
+        'associations': associations,
+        'attributes': attributes,
+        'methods': methods,
+        'custom_image_file': {'file': custom_image_file_path}
+    })
+    if cancel:
+        view.cancel.click()
+    else:
+        view.add.click()
+        view.flash.assert_no_error()
+
+    entity = self.instantiate(
+        name=name, description=description, attributes=attributes, associations=associations,
+        methods=methods, custom_image_file_path=custom_image_file_path
+    )
+    return entity
+
+
+@MiqImplementationContext.external_for(GenericObjectDefinition.update, ViaUI)
+def update(self, updates, reset=False, cancel=False):
+    """Update generic object definition
+
+    Args:
+        updates: dict, with fields to update
+    """
+    view = navigate_to(self, 'Edit')
+    view.fill({
+        'name': updates.get('name'),
+        'description': updates.get('description'),
+        'associations': updates.get('associations'),
+        'attributes': updates.get('attributes'),
+        'methods': updates.get('methods'),
+        'custom_image_file': {'file': updates.get('custom_image_file_path')}
+    })
+    if reset:
+        view.reset.click()
+    if cancel:
+        view.cancel.click()
+    elif not reset and not cancel:
+        view.save.click()
+        view = self.create_view(GenericObjectDefinitionDetailsView, override=updates)
+        assert view.is_displayed
+        view.flash.assert_no_error()
+
+
+@MiqImplementationContext.external_for(GenericObjectDefinition.delete, ViaUI)
+def delete(self):
+    """Delete generic object definition
+    """
+    view = navigate_to(self, 'Details')
+
+    view.configuration.item_select(
+        'Remove this Generic Object Classes from Inventory', handle_alert=True)
+    view = self.create_view(GenericObjectDefinitionAllView)
+    assert view.is_displayed
+    view.flash.assert_no_error()
+
+
+@MiqImplementationContext.external_for(GenericObjectDefinition.add_button, ViaUI)
+def add_button(self, name, description, image, request, button_type='Default', display=True,
+               dialog=None, open_url=None, display_for=None, submit_version=None,
+               system_message=None, attributes=None, role=None, button_group=None, cancel=False):
+    """Add button to generic object definition
+    """
```

Comment:
This needs detailed docblock describing these kwargs. `attributes` is a dict, `button_group` is just a name, not an object, what other strings are valid for `button_type` or where do those types come from (MIQ documentation?), whether `dialog` is an object, what's `request`?